zerver/actions/user_topics.py

```diff
@@ -40,5 +51,17 @@ def do_unmute_topic(user_profile: UserProfile, stream: Stream, topic: str) -> No
         remove_topic_mute(user_profile, stream.id, topic)
```

Comment:
there's a comment at the top of this function

```
# Note: If you add any new code to this function, the
# remove_topic_mute call in do_update_message will need to be
# updated for correctness.
```

And I do think you need to add something there. The function in question has a comment that says

```
                # Otherwise, we move the muted topic record for the user.
                # We call remove_topic_mute rather than do_unmute_topic to
                # avoid sending two events with new muted topics in
                # immediate succession; this is correct only because
                # muted_topics events always send the full set of topics.
```

`muted_topics` events always send the full set of topics, but that's not the case for `user_topics` events. So I think you do need to send a `user_topic_event` from there. Does that seem right?

Reply:
I've added the code that sends a `user_topic` event after the `remove_topic_mute` call. 
I was looking for tests in `test_events.py` for this piece of code and updating them accordingly. The tests currently do not seem to check this scenario, i.e. when a topic, muted by some users, is moved to another stream.
I'm not sure if there's a reason behind that. `test_send_stream_message_events` checks for events from `do_update_message` calls, but it doesn't check for `muted_topics` event that gets sent here.

Reply:
My guess is that this is an oversight: can you add an extra commit fixing that?

Reply:
Also, that duplication is really messy; I've added a bonus commit to just refactor to avoid duplicating the remove topic event code.